app/src/main/java/de/westnordost/streetcomplete/data/maptiles/MapTilesDownloader.kt

```diff
@@ -7,90 +7,90 @@ import de.westnordost.streetcomplete.ktx.format
+            bbox.enclosingTilesRect(zoom).asTilePosSequence().forEach { (x, y) ->
+                launch {
+                    try {
+                        val result = downloadTile(zoom, x, y)
+                        val logText = if (result.alreadyCached) "in cache" else "downloaded"
+                        Log.v(TAG, "Tile $zoom/$x/$y $logText")
+                        send(result)
+                    } catch (e: Exception) {
+                        Log.w(TAG, "Error retrieving tile $zoom/$x/$y: ${e.message}")
+                        send(DownloadFailure)
+                    }
+                }
+            }
         }
+    }
 
-    private suspend fun downloadTile(zoom: Int, x: Int, y: Int): DownloadResult = suspendCancellableCoroutine { cont ->
+    private suspend fun downloadTile(zoom: Int, x: Int, y: Int): DownloadSuccess = suspendCancellableCoroutine { cont ->
         /* adding trailing "&" because Tangram-ES also puts this at the end and the URL needs to be
            identical in order for the cache to work */
         val url = vectorTileProvider.getTileUrl(zoom, x, y) + "&"
         val httpUrl = HttpUrl.parse(url)
         require(httpUrl != null) { "Invalid URL: $url" }
 
-        val builder = Request.Builder()
+        val request = Request.Builder()
             .url(httpUrl)
             .cacheControl(cacheConfig.cacheControl)
-        builder.header("User-Agent", ApplicationConstants.USER_AGENT + " / " + Version.userAgent())
-        val call = okHttpClient.newCall(builder.build())
+            .header("User-Agent", ApplicationConstants.USER_AGENT + " / " + Version.userAgent())
+            .build()
+        val call = okHttpClient.newCall(request)
 
         /* since we use coroutines and this is in the background anyway, why not use call.execute()?
         *  Because we want to let the OkHttp dispatcher control how many HTTP requests are made in
         *  parallel */
         val callback = object : Callback {
             override fun onFailure(call: Call, e: IOException) {
-                Log.w(TAG, "Error retrieving tile $zoom/$x/$y: ${e.message}")
-                cont.resume(DownloadFailure)
+                cont.resumeWithException(e)
             }
 
             override fun onResponse(call: Call, response: Response) {
-                var size = 0
-                response.body()?.use { body ->
-                    // just get the bytes and let the cache magic do the rest...
-                    size = body.bytes().size
-                }
+                // just get the bytes and let the cache magic do the rest...
+                val size = response.body()?.use { it.bytes().size } ?: 0
                 val alreadyCached = response.cacheResponse() != null
-                val logText = if (alreadyCached) "in cache" else "downloaded"
-                Log.v(TAG, "Tile $zoom/$x/$y $logText")
                 cont.resume(DownloadSuccess(alreadyCached, size))
```

Comment:
`use` returns the inner return value, so we don't need to use a var here.

I went back and forth quite a bit on whether to replace this whole bit with `cont.resume(response) { response.close() }` and also move the other lines above. I ended up choosing this because it closes the response body as close to the source as possible — lower chance of accidental leaks.

Reply:
I'm also on the fence about whether the `DownloadResult` abstraction is needed at all — kotlin has a built in [`Result`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-result/) class we could use instead, which also includes the exception, in the case of failures, in case we wanted to move the logging all the way up to the top. But this seems good enough; I don't see any practical benefit to that except marginally cleaner code.